fix: [ANDROSDK-2194] Add hash-based unique database names

marcamsn opened the pull request:
This PR introduces hash-based database naming to guarantee uniqueness across accounts, even when server URLs normalize to the same base form.

**Unique Database Names**
- 	Adds an 8-char SHA-256 hash to database names: `processedUrl_username_hash_encrypted.db`
- 	Ensures no collisions between accounts pointing to similar URLs.
- 	Provides old format compatibility via getOldDatabaseName() for migration detection.

**New Migration301**
- 	Detecting accounts still using the old DB format.
- 	Renaming DB files.
- 	Migrating `sdk_resources` and `sdk_cache_resources` directories.
- 	Updating protectedDbName for pending imports.
- 	Updating stored configuration with new DB names.

marcamsn committed:
[ANDROSDK-2194] Add hash to database names for uniqueness

marcamsn committed:
[ANDROSDK-2194] Add tests

marcamsn committed:
[ANDROSDK-2194] Use `DB_SUFFIX` constant from `DatabaseNameGenerator`

marcamsn committed:
[ANDROSDK-2194] KtlintFormat and refactor code

marcamsn committed:
[ANDROSDK-2194] Add versioning to configuration migrations and refactor tests

marcamsn committed:
[ANDROSDK-2194] Move companion object back to the end of the class

marcamsn committed:
[ANDROSDK-2194] Remove commented out line in test file

marcamsn committed:
[ANDROSDK-2194] Use 'require' for failing migration checks

taridepaco commented on core/src/main/java/org/hisp/dhis/android/core/configuration/internal/migration/Migration301.kt:
```diff
@@ -0,0 +1,202 @@
+        if (alreadyMigrated) {
+            Log.d(TAG, "Database name hash migration already applied")
+            return
+        }
```
This guard clause returns if at least one db has the new name, but what about in there are other dbs with the old one? wouldn't the checking above be using all() instead of any()?

marcamsn replied:
I think that if at least one migration has executed, the app has already applied all earlier migrations. Given that `any()` and `all()` as migration validations do the same, I opted for any to improve performance. :)

taridepaco commented on core/src/main/java/org/hisp/dhis/android/core/configuration/internal/migration/Migration301.kt:
```diff
@@ -0,0 +1,202 @@
+        account: DatabaseAccount,
+        oldDbName: String,
+        newDbName: String,
+    ): org.hisp.dhis.android.core.configuration.internal.DatabaseAccountImport? {
```
move the import to the top, when searching it's easy to have them all together.

marcamsn committed:
[ANDROSDK-2194] Fix import

vgarciabnz commented on core/src/main/java/org/hisp/dhis/android/core/configuration/internal/migration/Migration301.kt:
```diff
@@ -0,0 +1,203 @@
+        }
+
+        val alreadyMigrated = configuration.accounts().any { account ->
+            account.databaseName().matches(Regex(".*_[0-9a-f]{8}_(encrypted|unencrypted)\\.db"))
```
I think this regex will match if the username is 8 chars long with characters in that range, for example `servername_abcdabcd_encrypted.db`.

Couldn't we just assume that the migration has to run if this method has been called? There is a check in the caller function that verifies that the existing code is less that 301.

vgarciabnz commented on core/src/test/java/org/hisp/dhis/android/core/configuration/internal/DatabaseNameGeneratorShould.kt:
```diff
@@ -38,42 +38,121 @@ class DatabaseNameGeneratorShould {
+        val nameHttps = generator.getDatabaseName(url, username, true)
+        val nameHttp = generator.getDatabaseName(urlHttp, username, true)
+
+        assertThat(nameHttps).isEqualTo(nameHttp)
```
Maybe we can discuss this point, I am not sure if instances with different protocols should be considered the same.

marcamsn committed:
[ANDROSDK-2194] Do not check if migration has been already applied in Migration301

marcamsn committed:
[ANDROSDK-2194] Extract ServerUrlNormalizer

marcamsn committed:
[ANDROSDK-2194] Merge remote-tracking branch 'origin/develop' into ANDROSDK-2194

# Conflicts:
#	core/src/main/java/org/hisp/dhis/android/core/configuration/internal/DatabaseConfigurationHelper.kt
#	core/src/test/java/org/hisp/dhis/android/core/configuration/internal/DatabasesConfigurationHelperShould.kt

marcamsn committed:
[ANDROSDK-2194] Fix test after merge

marcamsn committed:
[ANDROSDK-2194] Fix ServerUrlNormalizerShould syntax
